File: src/LTAdmin/Core/Result.cs
namespace LTAdmin.Core;

/// <summary>
/// Codes d'erreur stables utilisés par les services. Ils permettent aux écrans
/// d'afficher des messages adaptés sans analyser le texte des exceptions.
/// </summary>
public static class ErrorCodes
{
    public const string Validation = "VALIDATION";
    public const string BusinessRule = "GESTION";
    public const string NotFound = "INTROUVABLE";
    public const string UniqueViolation = "DOUBLON";
    public const string ForeignKeyViolation = "LIAISON";
    public const string RequiredField = "REQUIS";
    public const string DataTooLong = "TROP_LONG";
    public const string NumericOverflow = "DEPASSEMENT";
    public const string Locked = "VERROUILLE";
    public const string Authentication = "AUTHENTIFICATION";
    public const string Authorization = "HABILITATION";
    public const string Technical = "TECHNIQUE";
}

/// <summary>
/// Résultat d'une opération de service, sans valeur de retour.
/// Un échec contient toujours un message en français affichable à l'utilisateur.
/// </summary>
public class Result
{
    public bool IsSuccess { get; protected init; }
    public bool IsFailure => !IsSuccess;
    public string Message { get; protected init; } = string.Empty;
    public string? Code { get; protected init; }
    public IReadOnlyList<string> Errors { get; protected init; } = Array.Empty<string>();

    public static Result Ok(string message = "")
        => new() { IsSuccess = true, Message = message };

    public static Result Fail(string message, string? code = null, IEnumerable<string>? errors = null)
        => new()
        {
            IsSuccess = false,
            Message = message,
            Code = code ?? ErrorCodes.BusinessRule,
            Errors = errors is null ? Array.Empty<string>() : errors.ToArray()
        };

    public string FullMessage()
    {
        if (Errors.Count == 0) return Message;
        if (string.IsNullOrWhiteSpace(Message)) return string.Join(Environment.NewLine, Errors);
        return Message + Environment.NewLine + string.Join(Environment.NewLine, Errors);
    }
}

/// <summary>
/// Résultat d'une opération de service avec valeur de retour.
/// </summary>
/// <typeparam name="T">Type de la valeur produite en cas de succès.</typeparam>
public sealed class Result<T> : Result
{
    public T? Value { get; private init; }

    public static Result<T> Ok(T value, string message = "")
        => new() { IsSuccess = true, Value = value, Message = message };

    public static new Result<T> Fail(string message, string? code = null, IEnumerable<string>? errors = null)
        => new()
        {
            IsSuccess = false,
            Value = default,
            Message = message,
            Code = code ?? ErrorCodes.BusinessRule,
            Errors = errors is null ? Array.Empty<string>() : errors.ToArray()
        };
}

File: ltadmin/core/result.py
"""Résultats d'opérations et codes d'erreur stables.

Les services ne lèvent jamais d'exception vers l'interface pour les cas
gérés : ils retournent un ``Result`` (ou ``ResultValue``) portant un message
en français affichable et un code stable parmi :const:`ERROR_CODES`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, Iterable, List, Optional, TypeVar

T = TypeVar("T")

VALIDATION = "VALIDATION"
BUSINESS_RULE = "GESTION"
NOT_FOUND = "INTROUVABLE"
UNIQUE_VIOLATION = "DOUBLON"
FOREIGN_KEY_VIOLATION = "LIAISON"
REQUIRED_FIELD = "REQUIS"
DATA_TOO_LONG = "TROP_LONG"
NUMERIC_OVERFLOW = "DEPASSEMENT"
LOCKED = "VERROUILLE"
AUTHENTICATION = "AUTHENTIFICATION"
AUTHORIZATION = "HABILITATION"
TECHNICAL = "TECHNIQUE"


@dataclass
class Result:
    """Résultat d'une opération de service, sans valeur de retour."""

    is_success: bool = True
    message: str = ""
    code: Optional[str] = None
    errors: List[str] = field(default_factory=list)

    @property
    def is_failure(self) -> bool:
        return not self.is_success

    @property
    def success(self) -> bool:
        """Alias lisible (utilisé par l'interface)."""
        return self.is_success

    @classmethod
    def ok(cls, message: str = "") -> "Result":
        return cls(is_success=True, message=message, code=None, errors=[])

    @classmethod
    def fail(
        cls,
        message: str,
        code: Optional[str] = None,
        errors: Optional[Iterable[str]] = None,
    ) -> "Result":
        return cls(
            is_success=False,
            message=message,
            code=code or BUSINESS_RULE,
            errors=list(errors) if errors else [],
        )

    def full_message(self) -> str:
        if not self.errors:
            return self.message
        detail = "\n".join(self.errors)
        if not self.message.strip():
            return detail
        return self.message + "\n" + detail


@dataclass
class ResultValue(Result, Generic[T]):
    """Résultat d'une opération de service avec valeur de retour."""

    value: Any = None

    @classmethod
    def ok(cls, value: T, message: str = "") -> "ResultValue[T]":
        return cls(is_success=True, value=value, message=message, code=None, errors=[])

    @classmethod
    def fail(
        cls,
        message: str,
        code: Optional[str] = None,
        errors: Optional[Iterable[str]] = None,
    ) -> "ResultValue[T]":
        return cls(
            is_success=False,
            value=None,
            message=message,
            code=code or BUSINESS_RULE,
            errors=list(errors) if errors else [],
        )
